[css-values] Implement attr() fallback support

AlexsanderDamaceno opened the pull request:
#### fe663b06e8da7388f053104ca13de9ca48730212
<pre>
[css-values] Implement attr() fallback support
<a href="https://bugs.webkit.org/show_bug.cgi?id=280406">https://bugs.webkit.org/show_bug.cgi?id=280406</a>

Reviewed by Tim Nguyen.

Implement initial fallback support for attr(), refactor attr into a class to facilitate easier extension for additional features.

* Source/WebCore/Headers.cmake:
* Source/WebCore/SmartPointerExpectations/RefCntblBaseVirtualDtorExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/css/CSSAttrValue.cpp: Added.
(WebCore::CSSAttrValue::create):
(WebCore::CSSAttrValue::equals const):
(WebCore::CSSAttrValue::customCSSText const):
* Source/WebCore/css/CSSAttrValue.h: Added.
* Source/WebCore/css/CSSFontFeatureValue.h:
* Source/WebCore/css/CSSPrimitiveValue.cpp:
(WebCore::CSSPrimitiveValue::CSSPrimitiveValue):
(WebCore::CSSPrimitiveValue::~CSSPrimitiveValue):
(WebCore::CSSPrimitiveValue::create):
(WebCore::CSSPrimitiveValue::stringValue const):
(WebCore::CSSPrimitiveValue::serializeInternal const):
(WebCore::CSSPrimitiveValue::createAttr): Deleted.
* Source/WebCore/css/CSSPrimitiveValue.h:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValue.cpp:
(WebCore::CSSValue::visitDerived):
* Source/WebCore/css/CSSValue.h:
(WebCore::CSSValue::isAttrValue const):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeAttr):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueContent):
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/serialize-values-expected.txt:
* Source/WebCore/css/calc/CSSCalcTree+Simplification.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Copy.cpp:
(WebCore::CSSCalc::copy):
* Source/WebCore/css/calc/CSSCalcTree+Copy.h:
* Source/WebCore/css/CSSCustomPropertyValue.cpp:
(WebCore::CSSCustomPropertyValue::isSyntaxValueString const):
* Source/WebCore/css/CSSCustomPropertyValue.h:
* Source/WebCore/css/CSSPrimitiveValue.cpp:
(WebCore::CSSPrimitiveValue::stringValue const):
* Source/WebCore/css/CSSPrimitiveValue.h:
* Source/WebCore/css/values/CSSPrimitiveNumericTypes.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+TimingFunction.cpp:

Canonical link: <a href="https://commits.webkit.org/285849@main">https://commits.webkit.org/285849@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8eff3cf1e845fca8dcaab979c06a98b35859ce08

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/73876 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/53305 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/26687 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/78260 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/25114 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/75993 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/62438 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/1090 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/58135 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16446 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/76943 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48249 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63562 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38528 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/45063 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/21051 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/23447 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66617 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21399 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/79765 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/1193 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/616 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66413 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/1336 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63575 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65692 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16292 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9588 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/7768 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/1157 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/3907 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/1186 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/1173 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/1192 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -943,23 +943,39 @@ RefPtr<CSSValue> consumeCursor(CSSParserTokenRange& range, const CSSParserContex
 }
 
-RefPtr<CSSValue> consumeAttr(CSSParserTokenRange args, const CSSParserContext& context)
+RefPtr<CSSPrimitiveValue> consumeAttr(CSSParserTokenRange args, const CSSParserContext& context)
```
Leaving the return value as `RefPtr<CSSValue>` should be fine?

AlexsanderDamaceno replied:
@nt1m for unskip this imported/w3c/web-platform-tests/css/css-values/attr-notype-fallback.html, should I only remove this part [ ImageOnlyFailure ] ?

nt1m replied:
Just remove the whole line

nt1m commented on Source/WebCore/css/CSSAttrValue.h:
```diff
@@ -0,0 +1,65 @@
+public:
+
+    static Ref<CSSAttrValue> create(String ElementAttr,  String fallback , bool hasFallback = false);
+    const String elementattr() const { return m_elementattr; }
```
```suggestion
    const String attributeName() const { return m_attribute; }
```

I like attributeName because it matches the spec: https://www.w3.org/TR/css-values-5/#typedef-attr-name

nt1m commented on Source/WebCore/css/CSSAttrValue.h:
```diff
@@ -0,0 +1,65 @@
+private:
+    explicit CSSAttrValue(String && ElementAttr, String && fallback , bool hasFallback = false)
+        : CSSValue(ClassType::Attr)
+        , m_elementattr(WTFMove(ElementAttr))
```
```suggestion
        , m_attributeName(WTFMove(attributeName))
```

nt1m commented on Source/WebCore/css/CSSAttrValue.h:
```diff
@@ -0,0 +1,65 @@
+    }
+
+
+    String m_elementattr;
```
```suggestion
    String m_attributeName;
```

nt1m commented on Source/WebCore/css/CSSAttrValue.h:
```diff
@@ -0,0 +1,65 @@
+
+    String m_elementattr;
+    String m_fallback;
+    bool m_hasFallback;
```
```suggestion
    String m_fallback;
    bool m_hasFallback;
```

No need for `m_hasFallback`, we could just use a nullString if there is no fallback. And check it using `fallback().isNull()`

nt1m commented on Source/WebCore/css/CSSAttrValue.h:
```diff
@@ -0,0 +1,65 @@
+
+
+private:
+    explicit CSSAttrValue(String && ElementAttr, String && fallback , bool hasFallback = false)
```
```suggestion
    explicit CSSAttrValue(String&& attributeName, String&& fallback)
```

nt1m commented on Source/WebCore/css/CSSAttrValue.h:
```diff
@@ -0,0 +1,65 @@
+class CSSAttrValue final : public CSSValue {
+public:
+
+    static Ref<CSSAttrValue> create(String ElementAttr,  String fallback , bool hasFallback = false);
```
```suggestion
    static Ref<CSSAttrValue> create(String attributeName,  String fallback = nullString());
```

nt1m commented on Source/WebCore/css/CSSAttrValue.cpp:
```diff
@@ -0,0 +1,55 @@
+
+namespace WebCore {
+
+Ref<CSSAttrValue> CSSAttrValue::create(String elementattr, String fallback, bool hasFallback)
```
```suggestion
Ref<CSSAttrValue> CSSAttrValue::create(String attributeName, String fallback)
```

nt1m commented on Source/WebCore/css/CSSPrimitiveValue.cpp:
```diff
@@ -1287,14 +1296,15 @@ String CSSPrimitiveValue::stringValue() const
     case CSSUnitType::CSS_PROPERTY_ID:
         return nameString(m_value.propertyID);
+    case CSSUnitType::CSS_ATTR:
+        return m_value.attr->elementattr();
```
```suggestion
        return m_value.attr->attribute();
```

nt1m commented on Source/WebCore/style/StyleBuilderCustom.h:
```diff
@@ -1265,13 +1265,23 @@ inline void BuilderCustom::applyValueContent(BuilderState& builderState, CSSValu
             const_cast<RenderStyle&>(builderState.parentStyle()).setHasAttrContent();
-        QualifiedName attr(nullAtom(), primitiveValue.stringValue().impl(), nullAtom());
+
+        auto cssAttrvalue =  primitiveValue.cssAttrValue();
```
```suggestion
        auto value = primitiveValue.cssAttrValue();
```

nt1m commented on Source/WebCore/style/StyleBuilderCustom.h:
```diff
@@ -1265,13 +1265,23 @@ inline void BuilderCustom::applyValueContent(BuilderState& builderState, CSSValu
+            if (cssAttrvalue->hasFallback())
+                return cssAttrvalue->fallback().impl();
+
+            return emptyAtom();
```
```suggestion
            auto fallback = value->fallback();
            return !fallback.isNull() ? fallback.impl() : emptyAtom();
```

nt1m reviewed (commented):
This is great work, thanks! I mostly have naming nits.

Can you name your commit to: "[css-values] Implement `attr()` fallback support" ? (I find mostly the uppercase `A` a bit odd)

Can you unskip: https://github.com/WebKit/WebKit/blob/07a610a729e3aef47b09f92fa218a80938730363/LayoutTests/TestExpectations#L4096 ?

and also rebaseline `imported/w3c/web-platform-tests/css/cssom/serialize-values.html` ?

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -943,23 +943,39 @@ RefPtr<CSSValue> consumeCursor(CSSParserTokenRange& range, const CSSParserContex
 
-    if (!args.atEnd())
-        return nullptr;
+    if (args.size() && args.peek().type() == CommaToken)
```
```suggestion
    if (!args.atEnd() && args.peek().type() == CommaToken)
```

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -943,23 +943,39 @@ RefPtr<CSSValue> consumeCursor(CSSParserTokenRange& range, const CSSParserContex
+    String fallback;
+    bool hasFallback = false;
+
+    if (args.size()) {
```
You want to do this based on whether `consumeCommaIncludingWhitespace(args);` was successful or not.

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -943,23 +943,39 @@ RefPtr<CSSValue> consumeCursor(CSSParserTokenRange& range, const CSSParserContex
+            fallback = token.value().convertToASCIILowercaseAtom();
+        else
+            fallback = token.value().toAtomString();
+        hasFallback = true;
```
I just noticed the spec specifies the fallback to be a `<declaration-value>` rather than a string. I think you want to re-use `consumeDeclarationValue` here, and in `CSSAttrValue`, store the fallback as a `RefPtr<CSSValue>` rather than a String, where `nullptr` represents no fallback.

AlexsanderDamaceno replied:
you are right, thank you.

nt1m commented on Source/WebCore/Headers.cmake:
```diff
@@ -841,6 +841,8 @@ set(WebCore_PRIVATE_FRAMEWORK_HEADERS
     crypto/SerializedCryptoKeyWrap.h
     crypto/WrappedCryptoKey.h
 
+    css/CSSAnchorValue.h
```
This file was removed.

nt1m commented on Source/WebCore/css/CSSPrimitiveValue.cpp:
```diff
@@ -350,18 +350,37 @@ CSSPrimitiveValue::CSSPrimitiveValue(CSSUnresolvedColor unresolvedColor)
+    setPrimitiveUnitType(CSSUnitType::CSS_ANCHOR);
+    m_value.anchor = &value.leakRef();
+}
+
```
This was deleted in a previous commit: https://github.com/WebKit/WebKit/commit/ed4b5629b9e568df871ec607de28e35353d6df06, can you delete it?

nt1m commented on Source/WebCore/css/CSSPrimitiveValue.cpp:
```diff
@@ -350,18 +350,37 @@ CSSPrimitiveValue::CSSPrimitiveValue(CSSUnresolvedColor unresolvedColor)
         break;
+    case CSSUnitType::CSS_ANCHOR:
+        m_value.anchor->deref();
+        break;
```
ditto

nt1m commented on Source/WebCore/css/CSSPrimitiveValue.cpp:
```diff
@@ -573,9 +592,14 @@ Ref<CSSPrimitiveValue> CSSPrimitiveValue::create(CSSUnresolvedColor value)
-    return adoptRef(*new CSSPrimitiveValue(WTFMove(value), CSSUnitType::CSS_ATTR));
+    return adoptRef(*new CSSPrimitiveValue(WTFMove(value)));
+}
+
```
ditto

nt1m commented on Source/WebCore/css/CSSPrimitiveValue.h:
```diff
@@ -118,6 +121,9 @@ class CSSPrimitiveValue final : public CSSValue {
     static Ref<CSSPrimitiveValue> create(const Length&);
     static Ref<CSSPrimitiveValue> create(const Length&, const RenderStyle&);
     static Ref<CSSPrimitiveValue> create(Ref<CSSCalcValue>);
+    static Ref<CSSPrimitiveValue> create(Ref<CSSAnchorValue>);
```
same here

nt1m commented on Source/WebCore/css/CSSPrimitiveValue.h:
```diff
@@ -223,6 +226,8 @@ class CSSPrimitiveValue final : public CSSValue {
 
     WEBCORE_EXPORT String stringValue() const;
     const CSSCalcValue* cssCalcValue() const { return isCalculated() ? m_value.calc : nullptr; }
+    const CSSAnchorValue* cssAnchorValue() const { return isAnchor() ? m_value.anchor : nullptr; }
```
same here

nt1m commented on Source/WebCore/css/CSSPrimitiveValue.h:
```diff
@@ -254,6 +259,8 @@ class CSSPrimitiveValue final : public CSSValue {
     CSSPrimitiveValue(double, CSSUnitType);
     explicit CSSPrimitiveValue(Ref<CSSCalcValue>);
     explicit CSSPrimitiveValue(CSSUnresolvedColor);
+    explicit CSSPrimitiveValue(Ref<CSSAnchorValue>);
```
same here

nt1m commented on Source/WebCore/css/CSSPrimitiveValue.h:
```diff
@@ -307,6 +314,8 @@ class CSSPrimitiveValue final : public CSSValue {
         uint64_t colorAsInteger;
         const CSSUnresolvedColor* unresolvedColor;
         const CSSCalcValue* calc;
+        const CSSAnchorValue* anchor;
```
same

nt1m commented on Source/WebCore/css/CSSValue.cpp:
```diff
@@ -103,6 +104,10 @@ template<typename Visitor> constexpr decltype(auto) CSSValue::visitDerived(Visit
 {
     using enum CSSValue::ClassType;
     switch (m_classType) {
+    case Anchor:
```
same

nt1m commented on Source/WebCore/css/CSSValue.h:
```diff
@@ -73,6 +73,8 @@ class CSSValue {
 
     WEBCORE_EXPORT String cssText() const;
 
+    bool isAnchorValue() const { return m_classType == ClassType::Anchor; }
```
same

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -948,17 +948,27 @@ RefPtr<CSSValue> consumeAttr(CSSParserTokenRange args, const CSSParserContext& c
-        return nullptr;
+    if (!args.atEnd() && args.peek().type() == CommaToken) {
+        if (!consumeCommaIncludingWhitespace(args))
+            return nullptr;
```
```suggestion
    if (!args.atEnd() && !consumeCommaIncludingWhitespace(args)) {
        return nullptr;
```

I think this should be sufficient.

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -948,17 +948,27 @@ RefPtr<CSSValue> consumeAttr(CSSParserTokenRange args, const CSSParserContext& c
+    RefPtr<CSSValue> fallback = nullptr;
 
+    if (args.size())
+        fallback = consumeDeclarationValue(args, context);
```
Not sure the args.size() check is useful here, the previous check covers it.

AlexsanderDamaceno replied:
@nt1m you have reason

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -948,17 +948,27 @@ RefPtr<CSSValue> consumeAttr(CSSParserTokenRange args, const CSSParserContext& c
 
+    if (args.size())
+        fallback = consumeDeclarationValue(args, context);
+
```
You'll want:
```
    if (!args.atEnd())
         return nullptr;
```

to reject junk that is after the fallback.

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -948,17 +948,27 @@ RefPtr<CSSValue> consumeAttr(CSSParserTokenRange args, const CSSParserContext& c
+
+    RefPtr<CSSValue> fallback = nullptr;
+
+    fallback = consumeDeclarationValue(args, context);
```
```suggestion
    RefPtr fallback = consumeDeclarationValue(args, context);
```

nt1m commented on Source/WebCore/style/StyleBuilderCustom.h:
```diff
@@ -1258,13 +1258,26 @@ inline void BuilderCustom::applyValueContent(BuilderState& builderState, CSSValu
+
+            if (fallback != nullptr) {
+                auto fallbackValue = downcast<CSSCustomPropertyValue>(fallback)->customCSSText();
+                return downcast<CSSCustomPropertyValue>(fallback)->customCSSText().substring(1, fallbackValue.length() - 2).impl();
```
Just use `return fallback->cssText()`?

AlexsanderDamaceno replied:
@nt1m in this case when I use cssText it return the whole token value with quotation marks like this "value"

nt1m replied:
See my other comment, we should have used `CSSParserImpl::consumeDeclarationValue`

nt1m commented on Source/WebCore/css/CSSPrimitiveValue.h:
```diff
@@ -118,6 +120,8 @@ class CSSPrimitiveValue final : public CSSValue {
     static Ref<CSSPrimitiveValue> create(const Length&, const RenderStyle&);
     static Ref<CSSPrimitiveValue> create(Ref<CSSCalcValue>);
+    static Ref<CSSPrimitiveValue> create(Ref<CSSAttrValue>);
+
```
nit: remove extra blank line

nt1m commented on Source/WebCore/css/CSSPrimitiveValue.cpp:
```diff
@@ -1277,14 +1286,15 @@ String CSSPrimitiveValue::stringValue() const
     case CSSUnitType::CSS_PROPERTY_ID:
         return nameString(m_value.propertyID);
+    case CSSUnitType::CSS_ATTR:
+        return m_value.attr->attributeName();
```
```suggestion
    case CSSUnitType::CSS_ATTR:
        return m_value.attr->cssText();
```

nt1m commented on Source/WebCore/css/CSSPrimitiveValue.cpp:
```diff
@@ -1498,7 +1508,7 @@ ALWAYS_INLINE String CSSPrimitiveValue::serializeInternal() const
 
     case CSSUnitType::CSS_ATTR:
-        return makeString("attr("_s, m_value.string, ')');
+        return m_value.attr->customCSSText();
```
```suggestion
        return m_value.attr->cssText();
```

nt1m replied:
My previous comment still applies (using `cssText()` here)

nt1m commented on Source/WebCore/css/CSSAttrValue.h:
```diff
@@ -0,0 +1,62 @@
+        , m_fallback(WTFMove(fallback))
+    {
+    }
+
```
nit: remove one blank line here

nt1m commented on Source/WebCore/css/CSSAttrValue.h:
```diff
@@ -0,0 +1,62 @@
+    ~CSSAttrValue();
+    bool equals(const CSSAttrValue& other) const;
+    String customCSSText() const;
+
```
nit: remove one blank line here

nt1m commented on Source/WebCore/css/CSSAttrValue.h:
```diff
@@ -0,0 +1,62 @@
+
+class CSSAttrValue final : public CSSValue {
+public:
+
```
nit: remove this blank line

nt1m commented on Source/WebCore/css/CSSAttrValue.cpp:
```diff
@@ -0,0 +1,54 @@
+
+Ref<CSSAttrValue> CSSAttrValue::create(String elementattr, RefPtr<CSSValue>&& fallback)
+{
+    return adoptRef(*new CSSAttrValue(WTFMove(elementattr), WTFMove(fallback)));
```
```suggestion
Ref<CSSAttrValue> CSSAttrValue::create(String attributeName, RefPtr<CSSValue>&& fallback)
{
    return adoptRef(*new CSSAttrValue(WTFMove(attributeName), WTFMove(fallback)));
```

nt1m commented on Source/WebCore/css/CSSAttrValue.cpp:
```diff
@@ -0,0 +1,54 @@
+
+String CSSAttrValue::customCSSText() const
+{
+    return makeString("attr("_s, m_attributeName.impl(), ')');
```
This needs to serialize the fallback if provided (this is what some of the failures in css/cssom/serialize-values.html are about).

```suggestion
    return makeString(
        "attr("_s,
        m_attributeName.impl(),
        m_fallback ? ", "_s : ""_s,
        m_fallback ? m_fallback->cssText() : ""_s,
        ')'
    );
```

nt1m commented on Source/WebCore/css/CSSCustomPropertyValue.h:
```diff
@@ -99,6 +99,7 @@ class CSSCustomPropertyValue final : public CSSValue {
     bool containsCSSWideKeyword() const;
     bool isAnimatable() const;
 
+
```
nit: undo this change

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -948,17 +948,27 @@ RefPtr<CSSValue> consumeAttr(CSSParserTokenRange args, const CSSParserContext& c
+
+    RefPtr<CSSValue> fallback = nullptr;
+
+    fallback = consumeDeclarationValue(args, context);
```
Ah I figured out what's wrong here, you weren't using the correct method, I was thinking of `CSSParserImpl::consumeDeclarationValue`, not the method in  `CSSPropertyParserHelpers`. You'll also need to pass the `propertyID` argument in `consumeAttr` (the only call site should be using `CSSPropertyContent`, but the spec allows using `attr()` everywhere)

AlexsanderDamaceno replied:
@nt1m  Ok, saw my mistake, Thank you man :).

While looking  the CSSParserImpl::consumeDeclarationValue function, I saw that can not call it  because the function is not static, but there is this function here CSSParserImpl::parseValue, that call consumeDelcarationValue and it is static, could I use this one?

nt1m replied:
@AlexsanderDamaceno Can you use `CSSParser::parseSingleValue` for now? It won't work for shorthands, but should be good enough for your purposes, since `content` is a longhand anyway.

nt1m commented on Source/WebCore/css/CSSCustomPropertyValue.h:
I think you can undo all the changes in this file and the C++ file with my suggested change of using `CSSParserImpl::consumeDeclarationValue`

nt1m commented on Source/WebCore/css/CSSCustomPropertyValue.h:
These changes and the ones in the corresponding C++ file don't seem needed anymore

nt1m commented on Source/WebCore/css/CSSCustomPropertyValue.h:
```diff
@@ -88,8 +88,10 @@ class CSSCustomPropertyValue final : public CSSValue {
+
     String customCSSText() const;
 
+
```
Can you remove these lines?

nt1m commented on Source/WebCore/css/CSSPrimitiveValue.cpp:
```diff
@@ -1745,10 +1756,12 @@ bool CSSPrimitiveValue::addDerivedHash(Hasher& hasher) const
         add(hasher, String { m_value.string });
         break;
+    case CSSUnitType::CSS_ATTR:
+        add(hasher, String { m_value.attr->attributeName() });
```
You should probably also hash the fallback with it.

AlexsanderDamaceno replied:
Should call add on the fallback too? or use make string to concatenate and then hash it ?

nt1m replied:
I'd probably just do it like calc(), `add(hasher, m_value.attr);`

nt1m commented on Source/WebCore/style/StyleBuilderCustom.h:
```diff
@@ -1258,13 +1258,23 @@ inline void BuilderCustom::applyValueContent(BuilderState& builderState, CSSValu
+                if (fallback->isPrimitiveValue() && downcast<CSSPrimitiveValue>(fallback)->isString())
+                    return downcast<CSSPrimitiveValue>(fallback)->stringValue().impl();
+            }
+            return emptyAtom();
```
This equivalent code is shorter:

```suggestion
            auto fallback = dynamicDowncast<CSSPrimitiveValue>(value->fallback());
            return fallback ? fallback.stringValue().impl() : emptyAtom();
```
```

AlexsanderDamaceno replied:
Thank much better indeed.

AlexsanderDamaceno replied:
@nt1m in this case I should also  verify if fallback is a string value, if not  display nothing.

nt1m commented on Source/WebCore/css/CSSPrimitiveValue.cpp:
```diff
@@ -1641,9 +1651,10 @@ bool CSSPrimitiveValue::equals(const CSSPrimitiveValue& other) const
     case CSSUnitType::CSS_FONT_FAMILY:
         return equal(m_value.string, other.m_value.string);
+    case CSSUnitType::CSS_ATTR:
+        return equal(m_value.attr->cssText(), other.m_value.attr->cssText());
```
Comparing the cssText() seems wrong, I would define a `operator==` on CSSAttrValue (that's just the default one) and use that here.

AlexsanderDamaceno replied:
There is a equals function in CSSAtrr, could use this one?

nt1m commented on Source/WebCore/css/CSSAttrValue.cpp:
```diff
@@ -0,0 +1,60 @@
+}
+
+
+
```
please collapse this down to a single line

nt1m commented on Source/WebCore/css/CSSAttrValue.cpp:
```diff
@@ -0,0 +1,60 @@
+        fallback ? ", "_s : ""_s,
+        fallback ? fallback->cssText() : ""_s,
+        ')'
+    );
```
Just use m_fallback directly? no need to store it in a `RefPtr`

AlexsanderDamaceno replied:
in this case, use m_fallback make the  mac-safer-cpp fail.

nt1m replied:
Hmm ok, it can just be `RefPtr fallback = m_fallback;` then. `<CSSValue>` is not needed

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/cssom/serialize-values-expected.txt:
```diff
@@ -326,10 +326,10 @@ PASS content: attr(foo-bar)
+PASS content: attr(foo_bar, "fallback")
 FAIL content: attr(|bar, "fallback") assert_equals: content raw inline style declaration expected "attr(bar, \"fallback\")" but got ""
-FAIL content: attr(foo, "") assert_equals: content raw inline style declaration expected "attr(foo)" but got ""
+FAIL content: attr(foo, "") assert_equals: content raw inline style declaration expected "attr(foo)" but got "attr(foo, \"\")"
```
I wonder if this test is correct. While for `content`, `""` is indeed the default, it's not necessarily the case when you start using different types in `attr()`

cc @CanadaHonk @tursunova @tabatkins @fantasai @emilio

tabatkins replied:
That's [this test](https://github.com/web-platform-tests/wpt/blob/master/css/cssom/serialize-values.html#L109), right?

If so, then the test is correct. The default fallback only cares about the specified type (which defaults to `<string>`), not the context it's it.

nt1m commented on Source/WebCore/css/CSSAttrValue.cpp:
```diff
@@ -0,0 +1,63 @@
+        && fallback->cssText() == other.m_fallback->cssText();
+    }
+
+    return m_attributeName == other.m_attributeName;
```
```suggestion
    return m_attributeName == other.m_attributeName
        && m_fallback == other.m_fallback;
```

This whole function can just be that ^

No need to invoke `cssText()` nor null check

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -951,17 +951,29 @@ RefPtr<CSSValue> consumeAttr(CSSParserTokenRange args, const CSSParserContext& c
+    RefPtr<CSSValue> fallback = nullptr;
+    if (args.peek().type() == StringToken) {
+        token    = args.consumeIncludingWhitespace();
+        fallback = CSSPrimitiveValue::create(token.value().toString());
```
nit: we usually don't align `=` with spaces

nt1m reviewed (commented):
I wonder if @weinig or @anttijk have comments, but this looks good from my side

nt1m commented on Source/WebCore/css/CSSAttrValue.cpp:
```diff
@@ -0,0 +1,59 @@
+{
+    RefPtr fallback = m_fallback;
+    RefPtr otherFallback = other.m_fallback;
+    return m_attributeName == other.m_attributeName && fallback == otherFallback;
```
You can just compare directly the fallbacks without storing them, given the `==` operator should just be consulting trivial getters

emilio replied:
Totally drive-by, but pretty sure that comparing the fallback RefPtrs doesn't do what you want, as it only does a shallow (pointer) comparison: https://searchfox.org/wubkat/rev/1e937703de6af995ef2cd93867292c9da1d3b86c/Source/WTF/wtf/RefPtr.h#198-202

AlexsanderDamaceno replied:
@nt1m

> You can just compare directly the fallbacks without storing them, given the `==` operator should just be consulting trivial getters

Do like this give one more error in safer-cpp

nt1m replied:
Emilio is right, can we compare the fallback `stringValue()` ?

cdumez commented on Source/WebCore/css/CSSValue.h:
```diff
@@ -292,7 +294,7 @@ class CSSValue {
 
     // NOTE: This class is non-virtual for memory and performance reasons.
```
See this comment. This class intentionally doesn't have a virtual destructor. Ignore the bot that's telling you to make it virtual.

AlexsanderDamaceno replied:
Ok, got it, thanks.

AlexsanderDamaceno replied:
@nt1m

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -951,17 +951,29 @@ RefPtr<CSSValue> consumeAttr(CSSParserTokenRange args, const CSSParserContext& c
+    if (!args.atEnd() && !consumeCommaIncludingWhitespace(args))
+        return nullptr;
+
+    RefPtr<CSSValue> fallback = nullptr;
```
```suggestion
    RefPtr<CSSValue> fallback;
```

nt1m commented on Source/WebCore/css/CSSAttrValue.cpp:
```diff
@@ -0,0 +1,59 @@
+        "attr("_s,
+        m_attributeName.impl(),
+        fallback ? ", "_s : ""_s,
+        fallback ? fallback->cssText() : ""_s,
```
Can we avoid serializing the fallback if `fallback->stringValue()` is empty? That will fix one of the subtests

nt1m commented on Source/WebCore/css/CSSAttrValue.cpp:
```diff
@@ -0,0 +1,59 @@
+        "attr("_s,
+        m_attributeName.impl(),
+        fallback ? ", "_s : ""_s,
+        fallback && fallback->stringValue().length() > 0 ? fallback->stringValue() : ""_s,
```
```suggestion
        fallback && !fallback->stringValue().isEmpty() ? ", "_s : ""_s,
        fallback && !fallback->stringValue().isEmpty() ? fallback->cssText() : ""_s,
```

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/cssom/serialize-values-expected.txt:
```diff
@@ -326,10 +326,10 @@ PASS content: attr(foo-bar)
+PASS content: attr(foo_bar, "fallback")
 FAIL content: attr(|bar, "fallback") assert_equals: content raw inline style declaration expected "attr(bar, \"fallback\")" but got ""
-FAIL content: attr(foo, "") assert_equals: content raw inline style declaration expected "attr(foo)" but got ""
+FAIL content: attr(foo, "") assert_equals: content raw inline style declaration expected "attr(foo)" but got "attr(foo, \"\")"
```
Can you run this test with `--reset-results` again? It should be passing with the change I suggested

AlexsanderDamaceno replied:
sure.

nt1m commented on Source/WebCore/css/CSSAttrValue.cpp:
```diff
@@ -0,0 +1,62 @@
+        "attr("_s,
+        m_attributeName.impl(),
+        fallback && !fallback->stringValue().isEmpty() ? ", "_s : ""_s,
+        fallback && !fallback->stringValue().isEmpty() ? fallback->stringValue() : ""_s,
```
```suggestion
        fallback && !fallback->stringValue().isEmpty() ? fallback->cssText() : ""_s,
```

Should be `cssText()`, otherwise it will serialize the string without quotes.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/cssom/serialize-values-expected.txt:
```diff
@@ -326,10 +326,10 @@ PASS content: attr(foo-bar)
-FAIL content: attr(foo-bar, "fallback") assert_equals: content raw inline style declaration expected "attr(foo-bar, \"fallback\")" but got ""
-FAIL content: attr(foo_bar, "fallback") assert_equals: content raw inline style declaration expected "attr(foo_bar, \"fallback\")" but got ""
+FAIL content: attr(foo-bar, "fallback") assert_equals: content raw inline style declaration expected "attr(foo-bar, \"fallback\")" but got "attr(foo-bar, fallback)"
+FAIL content: attr(foo_bar, "fallback") assert_equals: content raw inline style declaration expected "attr(foo_bar, \"fallback\")" but got "attr(foo_bar, fallback)"
```
Can you re-run the test with `--reset-results` with my suggestion?

AlexsanderDamaceno replied:
@nt1m  updated.

nt1m commented on Source/WebCore/css/CSSAttrValue.cpp:
```diff
@@ -0,0 +1,62 @@
+    RefPtr otherFallback = dynamicDowncast<CSSPrimitiveValue>(other.m_fallback);
+
+    if (fallback && otherFallback)
+        return m_attributeName == other.m_attributeName && fallback->stringValue()  == otherFallback->stringValue();
```
```suggestion
        return m_attributeName == other.m_attributeName && fallback->stringValue() == otherFallback->stringValue();
```

AlexsanderDamaceno committed:
[css-values] Implement attr() fallback support
https://bugs.webkit.org/show_bug.cgi?id=280406

Reviewed by Tim Nguyen.

Implement initial fallback support for attr(), refactor attr into a class to facilitate easier extension for additional features.

* Source/WebCore/Headers.cmake:
* Source/WebCore/SmartPointerExpectations/RefCntblBaseVirtualDtorExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/css/CSSAttrValue.cpp: Added.
(WebCore::CSSAttrValue::create):
(WebCore::CSSAttrValue::equals const):
(WebCore::CSSAttrValue::customCSSText const):
* Source/WebCore/css/CSSAttrValue.h: Added.
* Source/WebCore/css/CSSFontFeatureValue.h:
* Source/WebCore/css/CSSPrimitiveValue.cpp:
(WebCore::CSSPrimitiveValue::CSSPrimitiveValue):
(WebCore::CSSPrimitiveValue::~CSSPrimitiveValue):
(WebCore::CSSPrimitiveValue::create):
(WebCore::CSSPrimitiveValue::stringValue const):
(WebCore::CSSPrimitiveValue::serializeInternal const):
(WebCore::CSSPrimitiveValue::createAttr): Deleted.
* Source/WebCore/css/CSSPrimitiveValue.h:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValue.cpp:
(WebCore::CSSValue::visitDerived):
* Source/WebCore/css/CSSValue.h:
(WebCore::CSSValue::isAttrValue const):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeAttr):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueContent):
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/serialize-values-expected.txt:
* Source/WebCore/css/calc/CSSCalcTree+Simplification.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Copy.cpp:
(WebCore::CSSCalc::copy):
* Source/WebCore/css/calc/CSSCalcTree+Copy.h:
* Source/WebCore/css/CSSCustomPropertyValue.cpp:
(WebCore::CSSCustomPropertyValue::isSyntaxValueString const):
* Source/WebCore/css/CSSCustomPropertyValue.h:
* Source/WebCore/css/CSSPrimitiveValue.cpp:
(WebCore::CSSPrimitiveValue::stringValue const):
* Source/WebCore/css/CSSPrimitiveValue.h:
* Source/WebCore/css/values/CSSPrimitiveNumericTypes.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+TimingFunction.cpp:

Canonical link: https://commits.webkit.org/285849@main

emilio commented on Source/WebCore/css/CSSAttrValue.cpp:
```diff
@@ -0,0 +1,62 @@
+
+    if (fallback && otherFallback)
+        return m_attributeName == other.m_attributeName && fallback->stringValue() == otherFallback->stringValue();
+    return m_attributeName == other.m_attributeName;
```
This is wrong too, fwiw. This makes `attr(foo)` and `attr(foo, "fallback")` compare equal, doesn't it @nt1m?

nt1m replied:
Oops yes, @AlexsanderDamaceno can you fix that in a separate commit?

AlexsanderDamaceno replied:
@nt1m sure

AlexsanderDamaceno replied:
@emilio Thank you, keen eyes :).
